### Exercise 1: Pandas

The first exercise should help you to deepen the experiences of Pandas that you gained in the course. It contains multiple sub-exercises for each topic.

#### a) Load data and get an overview

Load the data "student.csv" to a DataFrame and get an overview of the data. Display your results and explain your insights of the data

In [42]:
import pandas
students = pandas.read_csv(r"pandas_exercise_data/student.csv", delimiter=";")
students

ID school sex  age address famsize Pstatus  Medu  Fedu      Mjob  ...  \
0      1     GP   F   18       U     GT3       A     4     4   at_home  ...   
1      2     GP   F   17       U     GT3       T     1     1   at_home  ...   
2      3     GP   F   15       U     LE3       T     1     1   at_home  ...   
3      4     GP   F   15       U     GT3       T     4     2    health  ...   
4      5     GP   F   16       U     GT3       T     3     3     other  ...   
..   ...    ...  ..  ...     ...     ...     ...   ...   ...       ...  ...   
390  391     MS   M   20       U     LE3       A     2     2  services  ...   
391  392     MS   M   17       U     LE3       T     3     1  services  ...   
392  393     MS   M   21       R     GT3       T     1     1     other  ...   
393  394     MS   M   18       R     LE3       T     3     2  services  ...   
394  395     MS   M   19       U     LE3       T     1     1     other  ...   

    famrel freetime goout  Dalc  Walc  health absences  G1  G2  G3  
0        4        3     4     1     1       3        6   5   6   6  
1        5        3     3     1     1       3        4   5   5   6  
2        4        3     2     2     3       3       10   7   8  10  
3        3        2     2     1     1       5        2  15  14  15  
4        4        3     2     1     2       5        4   6  10  10  
..     ...      ...   ...   ...   ...     ...      ...  ..  ..  ..  
390      5        5     4     4     5       4       11   9   9   9  
391      2        4     5     3     4       2        3  14  16  16  
392      5        5     3     3     3       3        3  10   8   7  
393      4        4     1     3     4       5        0  11  12  10  
394      3        2     3     3     3       5        5   8   9   9  

[395 rows x 34 columns]

The file contents above display data of 395 students, each containing 35 columns. This includes basic data like, sex, age, address, as well as what I suppose to be the amount of days one spends on certain activities, like family, going out, drinking alcohol, exercising, etc.

#### b) Indexing and selection of data

Select all students, that have "at_home" as Mjob and "teacher" as Fjob. Additionaly get the studytime of those students.\
Select all students, that have zero failures. Are there more students that have zero failures than those who have at least one?\
Select all students with a uneven ID.\
Select all students that have an even amount of famrel.\
Is there any student, who is not paid, has no activities and no internet?\
\
Use loc, iloc, [] and/or direct attribute access

In [73]:
# 1)
athome_teaching = students.loc[lambda x: (x.Mjob == "at_home") & (x.Fjob == "teacher")].loc[:,"studytime"]
print(f"There are {len(athome_teaching)} students that study at home and work as a teacher.")

There are 2 students that study at home and work as a teacher.


In [46]:
# 2)
not_failed = students.loc[lambda x: x.failures == 0]
failed = students.loc[lambda x: x.failures > 0]

print(f"{len(failed)} students failed at least once, {len(not_failed)} never failed.")

83 students failed at least once, 312 never failed.


In [71]:
# 3)
id_uneven = students.loc[lambda x: x.ID % 2 != 0]
print(f"There are {len(id_uneven)} students with an uneven ID")

There are 198 students with an uneven ID


In [70]:
# 4)
famrel_even = students.loc[lambda x: x.famrel % 2 == 0]
print(f"There are {len(famrel_even)} students with an even famrel column")

There are 213 students with an even famrel column


In [67]:
# 5) Is there any student, who is not paid, has no activities and no internet?
nonono = students.loc[lambda x: (x.paid == "no") & (x.activities == "no") & (x.internet == "no")]
print(f"There are {len(nonono)} students who are not paid, have no activities and no internet.")

There are 22 students who are not paid, have no activities and no internet.


#### c) Merge, join and concatenate

Load the data "student_1.csv" and "student_2.csv" and combine the two datasets in a correct way. Save the new DataFrame to use it later.

In [105]:
st1 = pandas.read_csv(r"pandas_exercise_data/student_1.csv", delimiter=";")
st2 = pandas.read_csv(r"pandas_exercise_data/student_2.csv", delimiter=";")

In [103]:
st_combined = pandas.merge(st2,st1)
st_combined.to_csv(r"students_out.csv", sep=";", index=False)
st_combined

ID      Mjob      Fjob  reason guardian  traveltime  studytime  \
0      1   at_home   teacher  course   mother           2          2   
1      2   at_home     other  course   father           1          2   
2      3   at_home     other   other   mother           1          2   
3      4    health  services    home   mother           1          3   
4      5     other     other    home   father           1          2   
..   ...       ...       ...     ...      ...         ...        ...   
390  391  services  services  course    other           1          2   
391  392  services  services  course   mother           2          1   
392  393     other     other  course    other           1          1   
393  394  services     other  course   mother           3          1   
394  395     other   at_home  course   father           1          1   

     failures schoolsup famsup  ... famrel freetime goout Dalc Walc health  \
0           0       yes     no  ...      4        3     4    1    1      3   
1           0        no    yes  ...      5        3     3    1    1      3   
2           3       yes     no  ...      4        3     2    2    3      3   
3           0        no    yes  ...      3        2     2    1    1      5   
4           0        no    yes  ...      4        3     2    1    2      5   
..        ...       ...    ...  ...    ...      ...   ...  ...  ...    ...   
390         2        no    yes  ...      5        5     4    4    5      4   
391         0        no     no  ...      2        4     5    3    4      2   
392         3        no     no  ...      5        5     3    3    3      3   
393         0        no     no  ...      4        4     1    3    4      5   
394         0        no     no  ...      3        2     3    3    3      5   

     absences  G1  G2  G3  
0           6   5   6   6  
1           4   5   5   6  
2          10   7   8  10  
3           2  15  14  15  
4           4   6  10  10  
..        ...  ..  ..  ..  
390        11   9   9   9  
391         3  14  16  16  
392         3  10   8   7  
393         0  11  12  10  
394         5   8   9   9  

[395 rows x 26 columns]

#### d) Categorical Data and Grouping

Select three columns that are - in your opinion - categorical data and transform them to a categorical column. Rename the values of two of the columns in \
a) numerical values\
b) add a prefix

In [99]:
students.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 395 entries, 0 to 394
Data columns (total 34 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   ID          395 non-null    int64 
 1   school      395 non-null    object
 2   sex         395 non-null    object
 3   age         395 non-null    int64 
 4   address     395 non-null    object
 5   famsize     395 non-null    object
 6   Pstatus     395 non-null    object
 7   Medu        395 non-null    int64 
 8   Fedu        395 non-null    int64 
 9   Mjob        395 non-null    object
 10  Fjob        395 non-null    object
 11  reason      395 non-null    object
 12  guardian    395 non-null    object
 13  traveltime  395 non-null    int64 
 14  studytime   395 non-null    int64 
 15  failures    395 non-null    int64 
 16  schoolsup   395 non-null    object
 17  famsup      395 non-null    object
 18  paid        395 non-null    object
 19  activities  395 non-null    object
 20  nursery   